### PR TITLE
 [Form] Add FilterChoiceLoader + choice_filter option

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no
| Tests pass?   | yes    
| Fixed tickets | #11847, #14072, #28607, #28542
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/issues/10413

I did it :) i understand form choice lists.. took me only 2 days :joy: 

But here it is, a `choice_filter` option for the `ChoiceType`. Allowing to filter the list of choices independent from any choice loader / array.

This is no magic bullet. To truly optimize one should always prefer a filtered choice list (e.g. passing a query builder to the doctrine choice loader). But for built-in / pre-defined choice loaders, e.g. Intl, this is convenient.

Thanks to @sstok and @yceruto also.

- [x] agree on approach (per https://github.com/symfony/symfony/issues/14072#issuecomment-190107007)
- [x] consider delegating array choices to `ArrayChoiceList` for sanity :)
- [x] update filtering array choices (see todo in code)
- [x] fix callable argument consistency (`choice` only)
- [x] add form integration/rendering tests

### CODE REVIEW

File: src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        // Optimize
+        if (empty($values)) {
+            return array();
+        }
+        if (null !== $this->choiceList) {
+            return $this->loadChoiceList($value)->getChoicesForValues($values);
+        }
+
+        return array_filter($this->loader->loadChoicesForValues($values, $value), $this->filter, ARRAY_FILTER_USE_BOTH);
```
Comment (stof) [inline src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php]: will this pass the same signature than during loading ?

File: src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+     * {@inheritdoc}
+     */
+    public function loadValuesForChoices(array $choices, $value = null)
+    {
+        // Optimize
+        if (empty($choices)) {
+            return array();
+        }
+        if (null !== $this->choiceList) {
+            return $this->loadChoiceList($value)->getValuesForChoices($choices);
```
Comment (stof) [inline src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php]: can skip the call too

File: src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        return $this->loader->loadValuesForChoices(array_filter($choices, $this->filter, ARRAY_FILTER_USE_BOTH), $value);
```
Comment (stof) [inline src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php]: I'm quite sure this does not have the same arguments for the filter, as keys in this array have nothing to do with the original keys

**Comment by stof** [comment]
>     * fix callable argument consistency (`choice, key, value` ideally)

Looking at the places where you use the callable, you cannot do that, as you may only have the choice at that point.

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 4]
```diff
@@ -391,14 +405,24 @@ private function createChoiceList(array $options)
...
         if (null !== $options['choice_loader']) {
             return $this->choiceListFactory->createListFromLoader(
-                $options['choice_loader'],
+                null === $options['choice_filter'] ? $options['choice_loader'] : new FilterChoiceLoader($options['choice_loader'], $options['choice_filter']),
                 $options['choice_value']
             );
         }
 
         // Harden against NULL values (like in EntityType and ModelType)
         $choices = null !== $options['choices'] ? $options['choices'] : array();
```
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: what i like to do here is:

```php
$loader = new CallbackChoiceLoader(function () use ($choices) {
    return $choices;
});
```

and then proceed as above (like `choice_loader`)

File: src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php (hunk 1 above)
Comment (ro0NL) [reply src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php]: updated. now always provives `$choice[, $key]`

File: src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php (hunk 3 above)
Comment (ro0NL) [reply src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php]: you're right. forgot about original keys :( then `$choice` only :) which should be only needed for filtering anyway :+1:

File: src/Symfony/Component/Form/Tests/Fixtures/Author.php [hunk 5]
```diff
@@ -42,9 +42,11 @@ private function getPrivateGetter()
         return 'foobar';
     }
 
-    public function setAustralian($australian)
+    public function setAustralian($australian): self
```
Comment (ro0NL) [inline src/Symfony/Component/Form/Tests/Fixtures/Author.php]: @yceruto is this related?

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 6]
```diff
@@ -389,17 +404,18 @@ private function addSubForm(FormBuilderInterface $builder, string $name, ChoiceV
 
     private function createChoiceList(array $options)
     {
-        if (null !== $options['choice_loader']) {
```
[the remark this answers is not in the document]
Comment (ro0NL) [reply src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: done.

**Comment by ro0NL** [comment]
Ready :+1:

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 7]
```diff
@@ -296,6 +298,16 @@ public function configureOptions(OptionsResolver $resolver)
             return $choiceTranslationDomain;
         };
 
+        $choiceFilterNormalizer = function (Options $options, $choiceFilter) {
+            if (null !== $choiceFilter && !\is_callable($choiceFilter)) {
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: So if we pass a valid set of two values to filter out the list but the array happen to be callable this may break. Since it would add complexity to handle this edge case I suggest to either remove the support of array or document the limitation.

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 8]
```diff
@@ -391,15 +406,21 @@ private function createChoiceList(array $options)
...
             return $this->choiceListFactory->createListFromLoader(
-                $options['choice_loader'],
+                null === $options['choice_filter'] ? $options['choice_loader'] : new FilterChoiceLoader($options['choice_loader'], $options['choice_filter']),
                 $options['choice_value']
             );
         }
 
         // Harden against NULL values (like in EntityType and ModelType)
         $choices = null !== $options['choices'] ? $options['choices'] : array();
+        $loader = new CallbackChoiceLoader(function () use ($choices) {
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: This adds unnecessary overhead on every existing choice type in applications and prevents the caching factory decorator from reusing the same array of choices anymore since we always get a new hash for a new instance of loader, which is a big regression regarding performances.

Whereas we should be careful that all filtered lists can reuse the same cached choice list if possible in the first place (where I stopped last time IIRC).

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php (hunk 7 above)
Comment (ro0NL) [reply src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: not sure i understand

```php
$form = $this->createForm(ChoiceType::class, null, [
    'choices' => [1,2,3],
    'choice_filter' => [3],
]);
```

works as expected...

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php (hunk 8 above)
Comment (ro0NL) [reply src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: Hm i see, i didnt realize the hash is based on instans (spl_object_hash) for loaders.

What about storing these loaders in memory here, using `CachingFactoryDecorator::generateHash`. Or only normalize the array choices to a loader in case a filter is given.

**Comment by ro0NL** [comment]
@HeahDude i've optimized it a bit, now it only creates a new loader in case a filter is given. To me that's reasonable in terms of caching, the result would differ per filter anyway.. so no really cacheable to me.

**Comment by ro0NL** [comment]
to share cache between filtered/non-filtered choice lists we could add 

```
ChoiceLoaderFactoryInterface::createLoaderFromChoices(iterable $choices): ChoiceLoaderInterface
```

then we can always use a cacheable loader (either given or created from choices) before decorating it and passing it to `createListFromLoader`. But im not sure it's worth it...

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php (hunk 7 above)
Comment (Hanmac) [reply src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: @ro0NL   HashDude meant a choice_filter like this `["ClassName", "methodName"]` for this, is_callable might return true

Comment (ro0NL) [reply src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: i see, that's edgy yes :) :+1: for removing array support from me

**Comment by jo66** [comment]
@ro0NL choice_filter is nice option, will it be merged?

**Comment by ro0NL** [comment]
@jo66 rebased :) let's see if core wants to merge :+1:

**Comment by javiereguiluz** [comment]
Could we restart the work on this to finish it? It looks like a cool feature for Symfony 5.0!

**Comment by ro0NL** [comment]
I tend to close in favor of passing an optimized `choice_loader` (filtered and sorted as expected), that's the fastest way for core to move forward: no changes involved.

In #31593 i've also concluded this. 

The only "issue" for core is the built-in Intl choice types, which manage the choice loader for us.

Im thinking of a feature specific to Intl to handle the sorting/ordering here. Much like concluded in https://github.com/symfony/symfony/issues/32657#issuecomment-513833123 as well.

**Comment by stof** [comment]
@HeahDude didn't you start working on an alternative implementation at the choice factory level ? I cannot find a PR for it. You were supposed to open it as the result of the FOSSA Hackathon IIRC.

**Comment by yceruto** [comment]
@stof do you mean this one https://github.com/symfony/symfony/pull/30994 ?

**Comment by ro0NL** [comment]
closing this in favor of #32657 /  #30994
